src/main/kotlin/io/perfometer/runner/CoroutinesScenarioRunner.kt

```diff
@@ -47,18 +40,63 @@ internal class CoroutinesScenarioRunner(
     override suspend fun runStep(step: HttpStep) {
         when (step) {
             is RequestStep -> executeHttp(
-                httpClient(),
+                context().httpClient,
                 step,
             )
             is PauseStep -> pauseFor(step.duration)
+            is ParallelStep -> runParallel(step)
         }
     }
 
-    private suspend fun httpClient() =
-        coroutineContext[CoroutineHttpClient]?.httpClient ?: throw IllegalStateException()
+    /**
+     * Registers async step at the level of a 'user-coroutine'.
+     * Parallel blocks run inside one 'user-coroutine' share the same queue of async tasks.
+     */
+    override suspend fun registerAsync(step: HttpStep) = coroutineScope<Unit> {
+        context().asyncSteps.add(step)
+    }
+
+    /**
+     * Registers steps for async execution, calling step's asyncRegistrator function,
+     * and then iterates over registered tasks creating a separate coroutine for each.
+     *
+     * This method will not return until either all child coroutines are Completed, or terminate with an exception,
+     * or the parent 'user-coroutines' times out, effectively cancelling all pending jobs.
+     */
+    private suspend fun runParallel(step: ParallelStep) =
+        coroutineScope {
+            step.asyncRegistrator()
+            while (hasNextJob()) {
+                runJob(this@CoroutinesScenarioRunner, this)
+            }
+        }
+
+    private suspend fun runJob(
+        coroutinesScenarioRunner: CoroutinesScenarioRunner,
+        coroutineScope: CoroutineScope
+    ) {
+        val nextStep = coroutinesScenarioRunner.context().asyncSteps.poll()
+        coroutineScope.launch(CoroutineJobContext(httpClientFactory())) {
+            runStep(nextStep)
+        }
+    }
+
+    private suspend fun hasNextJob() = context().asyncSteps.size > 0
 
     private suspend fun pauseFor(duration: Duration) {
         delay(duration.toMillis())
         statistics.gather(PauseStatistics(duration))
     }
+
+    private suspend fun context(): CoroutineJobContext {
+        return coroutineContext[CoroutineJobContext.Key]
+            ?: throw IllegalStateException("Required context not setup")
+    }
+
+    private data class CoroutineJobContext(
+        val httpClient: HttpClient,
+        val asyncSteps: Deque<HttpStep> = ConcurrentLinkedDeque(),
```

Comment:
Is it possible that steps for single coroutine are accessed concurrently by multiple threads?

Reply:
I think no, at least not at the same time - concurrently. From what I've read though, coroutines are not thread bound, that is one coroutine started on thread X after suspending can be continued on a different underlying worker thread Y.